app/src/main/java/org/schabi/newpipe/player/mediasession/MediaSessionPlayerUi.java

```diff
@@ -0,0 +1,140 @@
+    }
+
+    @Override
+    public void onThumbnailLoaded(@Nullable final Bitmap bitmap) {
+        super.onThumbnailLoaded(bitmap);
+        if (sessionConnector != null) {
+            // the thumbnail is now loaded: invalidate the metadata to trigger a metadata update
+            sessionConnector.invalidateMediaSessionMetadata();
+        }
+    }
+
+
+    public void handleMediaButtonIntent(final Intent intent) {
+        MediaButtonReceiver.handleIntent(mediaSession, intent);
+    }
+
+    public Optional<MediaSessionCompat.Token> getSessionToken() {
+        return Optional.ofNullable(mediaSession).map(MediaSessionCompat::getSessionToken);
+    }
+
+
+    private ForwardingPlayer getForwardingPlayer() {
+        // ForwardingPlayer means that all media session actions called on this player are
+        // forwarded directly to the connected exoplayer, except for the overridden methods. So
+        // override play and pause since our player adds more functionality to them over exoplayer.
+        return new ForwardingPlayer(player.getExoPlayer()) {
+            @Override
+            public void play() {
+                player.play();
+                // hide the player controls even if the play command came from the media session
+                player.UIs().get(VideoPlayerUi.class).ifPresent(ui -> ui.hideControls(0, 0));
+            }
+
+            @Override
+            public void pause() {
+                player.pause();
+            }
+        };
+    }
+
+    private MediaMetadataCompat buildMediaMetadata() {
+        if (DEBUG) {
+            Log.d(TAG, "buildMediaMetadata called");
+        }
+
+        // set title and artist
+        final MediaMetadataCompat.Builder builder = new MediaMetadataCompat.Builder()
+                .putString(MediaMetadataCompat.METADATA_KEY_TITLE, player.getVideoTitle())
+                .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, player.getUploaderName());
+
+        // set duration (-1 for livestreams or if unknown, see the METADATA_KEY_DURATION docs)
+        final long duration = player.getCurrentStreamInfo()
+                .filter(info -> !StreamTypeUtil.isLiveStream(info.getStreamType()))
+                .map(info -> info.getDuration() * 1000L)
+                .orElse(-1L);
+        builder.putLong(MediaMetadataCompat.METADATA_KEY_DURATION, duration);
+
+        // set album art, unless the user asked not to, or there is no thumbnail available
+        final boolean showThumbnail = player.getPrefs().getBoolean(
+                context.getString(R.string.show_thumbnail_key), true);
+        Optional.ofNullable(player.getThumbnail())
+                .filter(bitmap -> showThumbnail)
+                .ifPresent(bitmap -> {
+                    builder.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, bitmap);
+                    builder.putBitmap(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON, bitmap);
+                });
```

Comment:
Wouldn't something like
```java
        if (player.getPrefs().getBoolean(context.getString(R.string.show_thumbnail_key), true)) {
            final BitMap bitmap = player.getThumbnail(); // might be done inline
            builder.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, bitmap);
            builder.putBitmap(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON, bitmap);
        }
```
be better here?

Reply:
[Previously](https://github.com/TeamNewPipe/NewPipe/blob/cacce6d2d0946ecf2b3cd79e20129031c2c0ca45/app/src/main/java/org/schabi/newpipe/player/helper/MediaSessionManager.java#L137) the bitmap was set only if it was non null, so I made sure the same behavior is kept, avoiding setting `null` thumbnails, even though it should not really be a problem.